runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/work/processing/StreamingWorkScheduler.java

```diff
@@ -0,0 +1,437 @@
+        .map(size -> (long) size)
+        .reduce(0L, Long::sum);
+  }
+
+  private static Windmill.WorkItemCommitRequest.Builder initializeOutputBuilder(
+      ByteString key, Windmill.WorkItem workItem) {
+    return Windmill.WorkItemCommitRequest.newBuilder()
+        .setKey(key)
+        .setShardingKey(workItem.getShardingKey())
+        .setWorkToken(workItem.getWorkToken())
+        .setCacheToken(workItem.getCacheToken());
+  }
+
+  private static Windmill.WorkItemCommitRequest buildWorkItemTruncationRequest(
+      ByteString key, Windmill.WorkItem workItem, int estimatedCommitSize) {
+    Windmill.WorkItemCommitRequest.Builder outputBuilder = initializeOutputBuilder(key, workItem);
+    outputBuilder.setExceedsMaxWorkItemCommitBytes(true);
+    outputBuilder.setEstimatedWorkItemCommitBytes(estimatedCommitSize);
+    return outputBuilder.build();
+  }
+
+  /** Sets the stage name and workId of the Thread executing the {@link Work} for logging. */
+  private static void setUpWorkLoggingContext(String workLatencyTrackingId, String computationId) {
+    DataflowWorkerLoggingMDC.setWorkId(workLatencyTrackingId);
+    DataflowWorkerLoggingMDC.setStageName(computationId);
+  }
+
+  private static String getShuffleTaskStepName(MapTask mapTask) {
+    // The MapTask instruction is ordered by dependencies, such that the first element is
+    // always going to be the shuffle task.
+    return mapTask.getInstructions().get(0).getName();
+  }
+
+  private static long computeSourceBytesProcessed(
+      DataflowWorkExecutor workExecutor, String sourceBytesCounterName) {
+    HashMap<String, ElementCounter> counters =
+        ((DataflowMapTaskExecutor) workExecutor)
+            .getReadOperation()
+            .receivers[0]
+            .getOutputCounters();
+
+    return Optional.ofNullable(counters.get(sourceBytesCounterName))
+        .map(counter -> ((OutputObjectAndByteCounter) counter).getByteCount().getAndReset())
+        .orElse(0L);
+  }
+
+  /** Resets logging context of the Thread executing the {@link Work} for logging. */
+  private void resetWorkLoggingContext(String workLatencyTrackingId) {
+    sampler.resetForWorkId(workLatencyTrackingId);
+    DataflowWorkerLoggingMDC.setWorkId(null);
+    DataflowWorkerLoggingMDC.setStageName(null);
+  }
+
+  /**
+   * Schedule work for execution. Work may be executed immediately, or queued and executed in the
+   * future. Only one work may be "active" (currently executing) per key at a time.
+   */
+  public void scheduleWork(
+      ComputationState computationState,
+      Windmill.WorkItem workItem,
+      Work.Watermarks watermarks,
+      Work.ProcessingContext.WithProcessWorkFn processingContext,
+      Collection<Windmill.LatencyAttribution> getWorkStreamLatencies) {
+    Work scheduledWork =
+        Work.create(
+            workItem,
+            watermarks,
+            processingContext.setProcessWorkFnAndBuild(work -> execute(computationState, work)),
```

Comment:
instead of binding the execute fn here, how about we bind the execution to computationstate as Consumer<Work> when constructing it?

Then I think that this method could just take Work, and work creation would be simpler since there wouldn't be this lazy binding.

Then computationState instead of scheduling work as  runnable directly would schedule that consumer bound with the work.

Reply:
because of the dependency graph made some modifications

but was able to remove this lazy binding and seperate Work from Consumer<Work>, combining them in ExecutableWork.java